### ISSUE TITLE
Cache doesn't break on new Calendar plugin event creation

### ISSUE THREAD

**Issue by MarcHartwig13** [putyourlightson/craft-blitz#131]
Craft CMS: 3.3.1.2
Calendar: 2.0.24
Blitz: 2.1.5

I'm running into an issue where when I add a new Solspace Calendar event the blitz plugin doesn't break the cache correctly. The specific page I'm looking to break is a JSON feed built with twig. Here is our normal process of adding an event:

- Create an entry with the event page's information.
- Go to the Calendar plugin and create an event in the plugin and associate it with the previously mentioned entry.

When saving the entry the cache breaks correctly. When saving the Calendar event blitz looks like it attempts to warm the cache but everything happens so quickly I can't tell if it actually does. To verify if the cache was rebuilt I went into my `web/cache/blitz` directory and watched the time stamp of when the folder was updated. When initially creating the Calendar event the cache doesn't break/rebuild.

If I go back to the entry and re-save it after I create the Calendar event the cache breaks/rebuilds correctly and is updated. After this if I go and update the Calendar event and save it then it also breaks/rebuilds the cache correctly. So it seems to be an issue only when a Calendar event is initially created.

I'm wondering if it has something to do with the loops I have in my twig code:

`{% for event in eventsActivities %}
  {% set eventDatesTimes = craft.calendar.events({ rangeStart: 'today', rangeEnd: 'today + 6 months' }).relatedTo({targetElement: event}).all() ?? null %}
  {% for dateTime in eventDatesTimes %}
    ...
  {% endfor %}
{% endfor %}`

eventsActivities are craft entries. eventDatesTimes are calendar events. I'm wondering if the cache doesn't break on the initial save of a calendar event because the loop for the calendar events is nested within the entry loop?

Thanks,
Marc

**Comment by putyourlightson** [comment]
Yes, it is very possible that the calendar events query embedded within the entries loop is the issue. Would it not make more sense to loop over the calendar events and fetch their entries using [eager loading](https://docs.craftcms.com/v3/dev/eager-loading-elements.html)? Besides being a more intuitive and much more performant way of doing things, it may also solve the issue you are experiencing with the cache not being cleared.

**Comment by MarcHartwig13** [comment]
Hi Ben,

Thanks for the tip! I've implemented the change you suggested and now the cache for the page isn't rebuilding on either the craft entry save or the Calendar Event save. I've narrowed it down to the initial save of either doesn't trigger Blitz to break the cache and rebuild it. If I manually `Refresh` the cache from the utilities and then make an update to either the craft entry or the Calendar Event it will update the cache as expected.

Here is my new twig code:
```
{% set eventDatesTimes = craft.calendar.events({ rangeStart: 'today', rangeEnd: 'today + 6 months' }).with(['pageInformation','pageInformation:featuredImage', 'pageInformation']).all() ?? null %}

{% for dateTime in eventDatesTimes %}
    {% set event = dateTime.pageInformation[0] %}
    ...
{% endfor %}
```

pageInformation is my entries field on my Calendar Event.

Thanks,
Marc

EDIT: Just thought of this, my Calendar Events do not have URI/URLs or a template associated with them. Could this have any effect? My feed URLs are also handle by a route and are not actually entries in the CMS but the route points to a twig file.

**Comment by putyourlightson** [comment]
After updating the template you should either delete the cached file manually or refresh the entire cache through the utility. It sounds like you did the latter and than now it is working, did I understand that correctly?

**Comment by MarcHartwig13** [comment]
After updating the template I did refresh the cache but it isn't working as expected. If I create a new craft entry or Calendar event after the cache refresh Blitz won't pick up on either of the new elements. If I also then edit those newly create elements and save them Blitz doesn't pick up on the change either. I have to manually refresh the cache for Blitz to be aware of any new content (for the feeds template). This is for the updated version of my code. If I edit an entry or Calendar Event that is already in the cached version of the feed then blitz breaks the cache and warms correctly. It seems to be an issue of when content is saved for the first time and seems to be related to the Calendar plugin. Sorry if I'm explaining this terribly.

It doesn't seem like the trigger referenced in this issue: https://github.com/putyourlightson/craft-blitz/issues/8 is working correctly with my current versions of Craft and the plugins. I could potentially create a button in the admin section of the CMS that triggers the cache to refresh. I checked my logs and didn't see anything related but I'd be happy to send those over if it'd help.

**Comment by MarcHartwig13** [comment]
Hey Ben,

This is super embarrassing for me but I jumped to step 5 instead of starting from the basics and working the problem through. The plugin is functioning correctly.

My issue is I need to update the feed cache _on save_ of Calendar Events, and two other entry channels. But thats not how Blitz works (Duh, Marc). So my new approach is to create a custom module that will hook into the Blitz service `refreshTaggedCache` I'll tag the feed page and then use the custom module to refresh that specific tag cache on save of the 3 elements/entries I outlined above.

Sorry for wasting your time. Appreciate your help.

Thanks,
Marc
